chi.go

```diff
@@ -1,81 +1,58 @@
 package chi
 
 import (
+	"context"
 	"net/http"
-
-	"golang.org/x/net/context"
 )
 
 // NewRouter returns a new Mux object that implements the Router interface.
 // It accepts an optional parent context.Context argument used by all
 // request contexts useful for signaling a server shutdown.
+
+// TODO: remove this parent context....? does it work..? do we need it?
+// or do we add our own signal to each routeContext ...?
+
 func NewRouter(parent ...context.Context) *Mux {
 	return NewMux(parent...)
 }
 
-// A Router consisting of the core routing methods used by chi's Mux.
-//
-// NOTE, the plan: hopefully once net/context makes it into the stdlib and
-// net/http supports a request context, we will remove the chi.Handler
-// interface, and the Router argument types will be http.Handler instead
-// of interface{}.
+// A Router consisting of the core routing methods used by chi's Mux,
+// built using just the stdlib.
 type Router interface {
 	http.Handler
-	Handler
-
-	Use(middlewares ...interface{})
-	Group(fn func(r Router)) Router
-	Route(pattern string, fn func(r Router)) Router
-	Mount(pattern string, handlers ...interface{})
-
-	Handle(pattern string, handlers ...interface{})
-	NotFound(h HandlerFunc)
 
-	Connect(pattern string, handlers ...interface{})
-	Head(pattern string, handlers ...interface{})
-	Get(pattern string, handlers ...interface{})
-	Post(pattern string, handlers ...interface{})
-	Put(pattern string, handlers ...interface{})
-	Patch(pattern string, handlers ...interface{})
-	Delete(pattern string, handlers ...interface{})
-	Trace(pattern string, handlers ...interface{})
-	Options(pattern string, handlers ...interface{})
+	Use(middlewares ...func(http.Handler) http.Handler)
+	Stack(fn func(r Router)) Router
+	Group(pattern string, fn func(r Router)) Router
+	Mount(pattern string, h http.Handler) // TODO: mount a Router instead of http.Handler?
+
+	Handle(pattern string, h http.Handler)
```

Comment:
Supporting this would also require a change to treeRouter.route to use string as key.

OR, which I believe is better

have a single route on `treeRouter` & have `map[method]http.Handler` on `node`

Benefits:
* single route tree, instead of multiple (you are adding chain to ALL method trees on .Route now)
* extensible method list
* proper HTTP 405 handling (we're returning 404 when there is a route, just not for current method, which is incorrect - 404 should be returned when there is no route for ANY method)
* no pointers in the tree -> no indirects -> routing tree as a single object and will often fit entirely in CPU cache (vs cache miss on every routing level - 2-5 CPU cycles vs 200-400 - see the very first table in https://blog.codinghorror.com/the-infinite-space-between-words/ )

, instead of having a `route[method]*treeRouter` in a tree